apps/aecore/lib/aecore/chain/chain_state.ex

```diff
@@ -5,43 +5,67 @@ defmodule Aecore.Chain.ChainState do
   """
 
   alias Aecore.Structures.SignedTx
+  alias Aecore.Structures.VotingTx
+  alias Aecore.Structures.SpendTx
   alias Aeutil.Serialization
   alias Aeutil.Bits
-
   require Logger
 
-  @type account_chainstate() ::
-          %{binary() =>
-            %{balance: integer(),
-              locked: [%{amount: integer(), block: integer()}],
-              nonce: integer()}}
+  @type account_chainstate() :: %{
+          binary() => %{
+            balance: integer(),
+            locked: [%{amount: integer(), block: integer()}],
+            nonce: integer()
+          }
+        }
 
-  @spec calculate_and_validate_chain_state!(list(), account_chainstate(), integer()) :: account_chainstate()
+  @spec calculate_and_validate_chain_state!(list(), account_chainstate(), integer()) ::
+          account_chainstate()
   def calculate_and_validate_chain_state!(txs, chain_state, block_height) do
     txs
-    |> Enum.reduce(chain_state, fn(transaction, chain_state) ->
+    |> Enum.reduce(chain_state, fn transaction, chain_state ->
       apply_transaction_on_state!(transaction, chain_state, block_height)
     end)
     |> update_chain_state_locked(block_height)
   end
 
-  @spec apply_transaction_on_state!(SignedTx.t(), account_chainstate(), integer()) :: account_chainstate()
+  @spec apply_transaction_on_state!(SignedTx.t(), map(), integer()) :: account_chainstate()
+  def apply_transaction_on_state!(
```

Comment:
this duplicates a lot of code, I think it would be better to pattern match in `apply_transaction_addition`